packaging/specifiers.py

```diff
@@ -146,7 +148,7 @@ def __eq__(self, other):
         elif not isinstance(other, self.__class__):
             return NotImplemented
 
-        return self._spec == other._spec
+        return hash(self) == hash(other)
```

Comment:
```suggestion
        return (self._spec[0], canonicalize_version(self._spec[1])) == (other._spec[0], canonicalize_version(other._spec[1]))
```
seems cleaner to me than relying on hashes.

Reply:
Personally I like to rely on `hash()`, since it would be easier to add entries to `_spec` in the future (if this is ever needed) without the risk of forgetting to update both of them.

Reply:
I agree with @uranusjr, I think this is actually cleaner and more clear. And since they're doing essentially the same thing, I don't see any issue with relying on `hash` here (although maybe I'm missing something).

Reply:
Well `hash()` computes a hash meaning you could have a hash collision between two completely different versions.

I'd find something like:
```
def _canonical_keys(self):
    return self._spec[0], canonicalize_version(self._spec[1])

def __eq__(self, other):
    return self._canonical_keys == other._canonical_keys

def __hash__(self):
    return hash(self._canonical_keys)
```
cleaner instead of relying on the huge hash space & hash randomization to avoid issues.